projects/baby_driver/scripts/test_gpio_interrupts.py

```diff
@@ -0,0 +1,342 @@
+        self.assertEqual(BABYDRIVER_DEVICE_ID, self.device_id)
+
+        register_gpio_interrupt(2, NUM_PINS_PER_PORT - 1, 1)
+        self.assertEqual(14, self.babydriver_id)
+        self.assertEqual([2, NUM_PINS_PER_PORT - 1, 1], self.data)
+        self.assertEqual(None, self.channel)
+        self.assertEqual(BABYDRIVER_CAN_MESSAGE_ID, self.msg_id)
+        self.assertEqual(BABYDRIVER_DEVICE_ID, self.device_id)
+
+        register_gpio_interrupt(2, 8, InterruptEdge.NUM_INTERRUPT_EDGES - 1)
+        self.assertEqual(14, self.babydriver_id)
+        self.assertEqual([2, 8, InterruptEdge.NUM_INTERRUPT_EDGES - 1], self.data)
+        self.assertEqual(None, self.channel)
+        self.assertEqual(BABYDRIVER_CAN_MESSAGE_ID, self.msg_id)
+        self.assertEqual(BABYDRIVER_DEVICE_ID, self.device_id)
+
+    @patch('can_util.send_message')
+    @patch('can_util.next_message')
+    def test_fail_conditions(self, mock_next_message, mock_send_message):
+        """Tests fail conditions"""
+
+        # Tests fail condition from register_gpio_interrupt parameter input
+        self.assertRaises(AttributeError, register_gpio_interrupt, 'G', 0, 0)
+        self.assertRaises(ValueError, register_gpio_interrupt, GpioPort.NUM_GPIO_PORTS, 4, 1)
+        self.assertRaises(ValueError, register_gpio_interrupt, -1, 9, 2)
+
+        self.assertRaises(ValueError, register_gpio_interrupt, GpioPort.B, NUM_PINS_PER_PORT, 0)
+        self.assertRaises(ValueError, register_gpio_interrupt, 'e', -1, 0)
+
+        self.assertRaises(ValueError, register_gpio_interrupt, GpioPort.D, 0, -1)
+        self.assertRaises(ValueError, register_gpio_interrupt, 'A', 15,
+                          InterruptEdge.NUM_INTERRUPT_EDGES)
+        self.assertRaises(AttributeError, register_gpio_interrupt, 'A', 15, 'no_edge')
+
+        self.assertRaises(ValueError, register_gpio_interrupt, 'A', 0, 0, 5)
+        self.assertRaises(ValueError, register_gpio_interrupt, 'A', 0, 0, "not_callable")
+
+        # Tests failing status code from can_util.next_message
+        mock_next_message.return_value.data = [0, 1]
+        self.assertRaises(Exception, register_gpio_interrupt, 'A', 0, 0)
+        self.assertRaises(Exception, register_gpio_interrupt, 'F', 15, 2)
+
+
+# pylint: disable=unused-argument
+class TestCallbackListener(unittest.TestCase):
+    """Test the callback_listener function"""
+
+    @patch('can_util.Message.from_msg')
+    # pylint: disable=no-self-use
+    def test_params_and_fail_conditions(self, mock_convert_can_msg):
+        """
+        Tests callback_listener by checking if it can trigger default/user-defined callbacks
+        upon receving can messages and tests fail conditions
+        """
+
+        def test_user_callback(info):
+            return f"Test callback, port:{info[0]}, pin:{info[1]}, edge:{info[2]}"
+
+        def incorrect_test_user_callback(port, pin, edge):
+            return f"Incorrect Test callback, {port}, {pin}, {edge}"
```

Comment:
Instead of returning values from the callbacks to test, why don't you set some properties on `self` like you do with the `parameter_test` mock above? That way you don't have to bend over backwards in your module to accommodate the tests.